### PR TITLE
Watch files opened via API

### LINKED ISSUE #467: Automatically reflect changes to SARIF files in the viewer
If a SARIF file is open in the viewer, any changes to that file should be automatically reflected.

Scenario:
1. Static analysis run produces a SARIF file; SARIF Viewer is opened
2. After fixing some of the issues, the static analysis tool is run again and updates/overwrites the existing SARIF file

Expected behavior: the SARIF viewer automatically reflects the changes in the file.

Actual behavior: nothing changes :(.

### PR DESCRIPTION
**Pull request by 50Wliu**
Implements & closes #467.

This is a very simple file-watching implementation - for every log opened via `api.openLogs()`, we'll add it to a file watcher. When those files change (or are deleted), we'll update the SARIF Viewer as appropriate. We'll remove the files from the watcher once they're closed.

What I don't know is how chokidar handles multiple `add` calls for the same file - will it start watching the same file _twice_? Or is it smart enough to know that it's already watching and no-op? I don't have an easy way to test that scenario, but I'm hoping it's a pretty rare edge case for someone to repeatedly `openLogs` on a file that's already open.

I've tested this locally and confirmed that the openLogs implementation works, but not that closeLogs successfully removes from the watcher (though I'd be very surprised if it didn't).

### CODE REVIEW

File: src/extension/index.ts
Comment (50Wliu) [inline src/extension/index.ts:74]: Not that it matters since we start out with an empty array, but `ignoreInitial` ensures that the change event doesn't get called when the watcher does the first pass on all the files to ready itself.

**Comment by 50Wliu** [comment]
> how often do you anticipate quick successions of edits/changes to the same file?

For our use case - never. I can see how difficult it'd be to add a debounce timeout for each file though.

> Note: I did a quick test and Chokidar seems to ignore adding files that have already been added.

Thanks!

> Stylistically

This makes sense. I was thinking it'd be nice to rely on the `initial` functionality. If I get what you're saying - all the API methods would act on `watchedLogs`, which internally would update `logs`. Everything else continues to use `logs` as usual.

Thanks for the quick turnaround! I'll work on those changes.

**Comment by jeffersonking** [comment]
> For our use case - never. I can see how difficult it'd be to add a debounce timeout for each file though.

Glad that's the case. Prefer to defer the debounce until absolutely needed.

**Comment by chrishuynhc** [comment]
Hi @jeffersonking, have you had the chance to finish reviewing? @50Wliu is OOF and asked me to follow up on this change :)

**Comment by jeffersonking** [comment]
@chrishuynhc Not yet. Will do today. Thanks.

**Comment by jeffersonking** [comment]
@50Wliu Good observations and I agree.

Ideally we would be able to add/remove watched logs (say from other code paths) without going the API code path. However, we can do that refactoring when that day comes.

File: src/extension/index.ts
Comment (jeffersonking) [inline src/extension/index.ts:86]: Should `watcher.add` happen after `store.logs.push`?

**Review by jeffersonking — APPROVED** [review]
Approved, pending the remaining comments.

File: src/extension/index.ts
Comment (50Wliu) [reply src/extension/index.ts:204]: My local TypeScript version was throwing red errors on this line. The changes aren't required, but I figure they at some point will be when this repo's TS is upgraded. Up to you if you want me to revert the changes, I'm fine either way.

File: src/extension/index.ts
Comment (50Wliu) [reply src/extension/index.ts:86]: Hmm. I think that makes sense in case the following happens:
1. We open a log
2. The log is added to the watcher
3. For whatever reason, loadLogs takes forever to finish
4. The log is changed on the filesystem after the watcher starts watching the file but before loadLogs finishes
5. Race condition?

Wouldn't have thought about this otherwise, thanks for pointing it out :)

Comment (50Wliu) [reply src/extension/index.ts:86]: Actually, turns out that loadLogs isn't even async, so we don't need to worry about that!
...which is a good thing, because my comment made me realize the same race conditions applied to multiple "change" events happening in short succession.

File: src/extension/index.ts
Comment (jeffersonking) [reply src/extension/index.ts:204]: What was the TS error saying? I'm not seeing the same.

Comment (50Wliu) [reply src/extension/index.ts:205]:
```
This expression is not callable.
  Each member of the union type '{ (callbackfn: (previousValue: string, currentValue: string, currentIndex: number, array: string[]) => string): string; (callbackfn: (previousValue: string, currentValue: string, currentIndex: number, array: string[]) => string, initialValue: string): string; <U>(callbackfn: (previousValue: U, currentValue: string, ...' has signatures, but none of those signatures are compatible with each other.ts(2349)
```
image

Which, after looking it up, seems to be because lines has a type of `RegExpMatchArray | []`, but `.reduce` has to act on a specific type of array.

File: src/extension/index.ts
Comment (jeffersonking) [reply src/extension/index.ts:86]: Yes, `loadLogs` is labeled as `async` but doesn't actually `await` anything. In the past we used to await VS Code prompts.

However, if two "change" events come in short succession, it is still possible to race, correct? There's nothing preventing two `loadLogs` calls running at the same time.

File: src/extension/index.ts
Comment (jeffersonking) [reply src/extension/index.ts:205]: That's a good reason. Let's keep your change.

File: src/extension/index.ts
Comment (50Wliu) [reply src/extension/index.ts:86]: Right. Let me think of that and come back to you with a few ideas.

File: src/extension/loadLogs.ts
Comment (jeffersonking) [inline src/extension/loadLogs.ts:14]: Prefer to defer housekeeping to another PR. There are other calls to `loadLogs` you have not fixed up consistently. But to fix all of them detracts from the clarity of this PR.

File: src/extension/index.ts
Comment (jeffersonking) [reply src/extension/index.ts:86]: Since we discussed don't anticipate short succession changes, I'd say a code comment acknowledging the possibility of a race is sufficient.

Comment (50Wliu) [reply src/extension/index.ts:93]: Sounds good, comment added. Thinking about it - I think if loadLogs is sync, the chance the race condition will actually occur will be extremely small if not impossible - you're banking on another change event getting fired _after_ the removal but _before_ the push.
I'm not familiar with JavaScript multithreading as I'd like to be, but if the extension can only process one "change" event at a time like I suspect - that makes it race-free.

**Comment by 50Wliu** [comment]
@jeffersonking, @EasyRhinoMSFT - any chance I could get a signoff + merge? I believe all concerns have been addressed :).

**Comment by 50Wliu** [comment]
Thank you! ❤️

**Comment by EasyRhinoMSFT** [comment]
Thank *you* Winston! This is an important feature and you've helped our goal of keeping the two Viewer extensions in sync. 😎

**Comment by 50Wliu** [comment]
@EasyRhinoMSFT - would you mind publishing a new release for this, please?

**Comment by EasyRhinoMSFT** [comment]
Sure thing -- apologies for forgetting this, I was on a two-week vacation.

**Comment by 50Wliu** [comment]
Thanks so much!

**Comment by EasyRhinoMSFT** [comment]
@50Wliu release to the Marketplace complete 👍

**Pull request merged**
